pkg/manager/impl/reservation_manager.go

```diff
@@ -0,0 +1,188 @@
+			reservationScope,
+		),
+		makeReservationFailure: labeled.NewCounter(
+			"make_reservation_failure",
+			"Number of times we failed to make reservation",
+			reservationScope,
+		),
+		getTagFailure: labeled.NewCounter(
+			"get_tag_failure",
+			"Number of times we failed to get tag",
+			reservationScope,
+		),
+	}
+
+	return &reservationManager{
+		repo:               repo,
+		reservationTimeout: reservationTimeout,
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's reserve a spot to work on
+			// generating the artifact.
+			status, err := r.makeReservation(ctx, request)
+			if err != nil {
+				r.systemMetrics.makeReservationFailure.Inc(ctx)
+				return nil, err
+			}
+
+			return &datacatalog.GetOrReserveArtifactResponse{
+				Value: &datacatalog.GetOrReserveArtifactResponse_ReservationStatus{
+					ReservationStatus: &status,
+				},
+			}, nil
+		}
+		logger.Errorf(ctx, "Failed retrieve tag: %+v, err: %v", tagKey, err)
+		r.systemMetrics.getTagFailure.Inc(ctx)
+		return nil, err
+	}
+
+	artifact, err := transformers.FromArtifactModel(tag.Artifact)
+	if err != nil {
+		return nil, err
+	}
+
+	return &datacatalog.GetOrReserveArtifactResponse{
+		Value: &datacatalog.GetOrReserveArtifactResponse_Artifact{
+			Artifact: artifact,
+		},
+	}, nil
+}
+
+func (r *reservationManager) makeReservation(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (datacatalog.ReservationStatus, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.ToReservationKey(*request.DatasetId, request.TagName)
+	rsv, err := repo.Get(ctx, reservationKey)
+
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Reservation does not exist yet so let's create one
+			err := repo.Create(ctx, models.Reservation{
+				ReservationKey: reservationKey,
+				OwnerID:        request.OwnerId,
+				ExpireAt:       r.now().Add(r.reservationTimeout),
+			})
+
+			if err != nil {
+				logger.Errorf(ctx, "Failed to create reservation: %+v, err %v", reservationKey, err)
+
+				return datacatalog.ReservationStatus{}, err
+			}
+
+			r.systemMetrics.reservationAcquiredViaCreate.Inc(ctx)
+
+			return datacatalog.ReservationStatus{
+				State:   datacatalog.ReservationStatus_ACQUIRED,
+				OwnerId: request.OwnerId,
+			}, nil
+		}
+		return datacatalog.ReservationStatus{}, err
+	}
+
+	now := r.now()
+	// Reservation already exists so there is a task already working on it
+	// Let's check if the reservation is expired.
+	if rsv.ExpireAt.Before(now) {
```

Comment:
this is weird too. I think we should make this entire thing one single request

GetOrMake(pass the time)
// in db psuedo code
```python
if record is none or record is expired:
   create new record
   return
return record
```

Reply:
So are you propose is to merge all the GET reservation, Create reservation and Update reservation db calls into a single Upsert request and do a GET in the end to retrieve the owner ID (necessary because we want to know who actually owns the reservation)

The reason for the separate API approach is that a GET RESERVATION call is needed to retrieve the owner of the reservation (possibly other task owns the reservation). Since we are doing a GET at first, why not just do the expiration time check at application layer so that we can avoid unnecessary db calls.

Reply:
The reason for the separate API approach is that GET RESERVATION call is always needed to retrieve the owner of the reservation . Since we are doing a GET at first, why not just do the expiration time check at application layer so that we can avoid unnecessary db calls.

Reply:
I agree that the get can be optimized, the only thing is if Get fails, you could still end up in a race. So, inside the i failure, we should do a conditional write and get or a Compare-and-swap operation. If we cannot do that, then maybe just do the upsert part atleast

Reply:
> I agree that the get can be optimized, the only thing is if Get fails, you could still end up in a race. So, inside the i failure, we should do a conditional write and get or a Compare-and-swap operation

Can you elaborate a bit on this?